### ISSUE TITLE
Outlier case in Excess Property detection in Union (a case that is -more- strict than it should be)

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#44856]
# Bug Report

### 🔎 Search Terms

Excess Property Checks

(1) 

From [Release Notes 3.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-5.html#improved-excess-property-checks-in-union-types)

> In TypeScript 3.5, the type-checker at least verifies that all the provided properties belong to some union member and have the appropriate type, meaning that the sample above correctly issues an error. Note that partial overlap is still permitted as long as the property types are valid.

### 🕗 Version & Regression Information
3.9.7 - 4.3.4 checked in playground to be the same.

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?target=7&ts=4.3.4#code/PTAEBUAsFNWgPADtAxgF2gE1AJwK4A2sAZgPY6iI6nI5oCW0AzqKcaHgHb2mehoBPZKAAUAHwCUoei1AAoEKICMUgLZ4maUCl5oAhvT56CBXNACOeejiyVqtBs1bs9Wons2tOsLjz6roVQAjaBwFMBEAJjU9AW1dAyNMTHoGXmM7GlDHFjRIV1AQgl4Ac35SUD1OON9eUADg0Lk5AG85UH4hWABBSNAAXlAWvQAuUE0cQxKAbkKxiamAX2n2zuFugGYBodHxtEnOGbnC0lJ3TlmUMaDT8+XVwXWAFm3h+f2p2aDr2+gqy7GnDwjRw91WbQ6HUU4AAEgBJADKoEREBhAFFQABVAByaIAGgAFNEAYXAaIAIqAAPKY8AAGThaIASqsOjpOJ5vqBemJNq9dgByAUAGmOQtFV34+Ggy1AikAMuSVJhMeglTgBThaBboAhxFAwFAAaywooQ+z0mUQoGI1FU3K2nFIWmMxQA7lhVotway5WAAEJomHdABqcOxAHFuUj8UTSRSfezOWMeZteS9Bm9QOKxSLtGN9ngZbNFIAUciVKrVGrQLBdpFdcHg5stLDYrDyoQ43Dqj2YnrkXqAA)

### 💻 Code

```
// The expected rule for properties of union type (|) is  
// (1) must contain all required properties of at least one union member
// (2) may contain additional properties that belong to any union member

{
  type A2 = {a: string; b: string};
  type A3 = {a: string; b: boolean; c: boolean};
  type A4 = {a: string; b: boolean; c: number};

  {
    // THIS IS THE UNEXPECTED OUTLIER (compiler error happens unexpectedly)
    const b: A2|A3 = {a: '', b: '', c: true}; // ❌ assignment strictly checked, extra prop from A3 not allowed
  }
  {
    // BEHAVING AS EXPECTED (no compiler error)
    const b: A2|A3|A4 = {a: '', b: '', c: true}; // ✔ assignments allow extra props of other union types
  }
}
```

### 🙁 Actual behavior

The line item marked `// THIS IS THE UNEXPECTED OUTLIER` is rejected as invalid for `A2|A3`
even though 
`{a:''",b:""}` contains all properties to satisfy `A2`, and `c:true` should be allowed by `A3`.

The oddity of that is accentuated by noting that the same value can be assigned to `A2|A3|A4`, where `A4` differs only from `A3` in having property-type `c:number` instead of `c:boolean`.

### 🙂 Expected behavior

The unexpected rejection of property `c` should be accepted as it is in the counterexample.

---

### TL;DR :  About the actual status quo limited type checking introduced with ts 3.5

From [Release Notes 3.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-5.html#improved-excess-property-checks-in-union-types)

> In TypeScript 3.5, the type-checker at least verifies that all the provided properties belong to some union member and have the appropriate type, meaning that the sample above correctly issues an error. Note that partial overlap is still permitted as long as the property types are valid.

That description does not actually match the observed status quo.  For example, that rule would imply that an empty object could always be assigned to a union of object with required properties, and that is obviously false.  So surely what the author meant was 

>    ... the type-checker at least verifies that all the required properties of some union member are provided ...

i.e., the rule for a set of properties to satisfy a union type (|) is  
- (1) must contain all required properties of at least one union member
- (2) may contain additional properties that belong to any union member

This actual observed behavior is stronger excess property checking than the original verbiage.

That would still make the outlier behavior in this bug report a bug.  It's just worth mentioning while we are on the topic.  I have create a separate issue for this #44871.

**Comment by MartinJohns** [comment]
See #20863.

**Comment by craigphicks** [comment]
> See #20863.

@MartinJohns 

That seems to be a post requesting a "Strict Union" - a topic I am familiar with that topic and understand and accept that typescript Union is NOT strict.

What I have found in practice is that this following rule is (almost) always satisfied: 

The rule for a property set to satisfy a union type (|) is
-  (1) must contain all required properties of at least one union member
-  (2) may contain additional properties that belong to any union member

The following quote from the [rulebook](https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#improved-excess-property-checks-in-union-types)

> In TypeScript 3.5, the type-checker at least verifies that all the provided properties belong to some union member and have the appropriate type, meaning that the sample above correctly issues an error. Note that partial overlap is still permitted as long as the property types are valid.

matches the rule I gave with a minor change -  

> ... the type-checker at least verifies that all the properties of some union member are provided ...

and I think that is what the writer intended because that is much closer to the actual behavior than what the write wrote. (What the writer actually wrote implies that an empty object should always be assignable to a union of property bearing objects, and that is obviously false.)

I am not interested in proposing a strict union or any major changes, but I do believe the intended behavior is supposed to be simple to understand and predictable, and that must be the rule I wrote above.  Therefore the outlier example I showed is a bug.

**Comment by MartinJohns** [comment]
> That seems to be a post requesting a "Strict Union"

Not really. It's about excess property checks, and excess property checks can not prevent the passing of additional properties (aka "strict").

**Comment by craigphicks** [comment]
@MartinJohns  -   

> "It's about excess property checks, and excess property checks can not prevent the passing of additional properties (aka "strict").

Agreed.  And the status of #20863 is "needs proposal" and "suggestion", specifically a proposal to change or add to the status quo to enable **stronger** excess property checking. That is fine - but does not overlap with this post.

In contrast, this bug report accepts the apparent status quo of limited excess-property detection behavior which came with typescript 3.5 (\*), and points out an unusual outlier case to that status which appears to be a bug.  That outlier case displays unexpected strict type checking, that doesn't follow the status quo of limited type checking. 

I am afraid you read the title of this question, saw the phrase "Excess Property detection", and assumed this post must belong as a comment under #20863.  I would very much appreciate it if you would look closer.  I will modify the title to improve clarity.

**Comment by sandersn** [comment]
Fixing this is quite tricky because the implementation is written cleverly and for speed. Plus, the semantics of excess property checks are so quirky that changing it is very likely to break other code.

**Comment by craigphicks** [comment]
> Plus, the semantics of excess property checks are so quirky that changing it is very likely to break other code.

The bug is on the side of being too strict.  Fixing the bug would only remove compiler errors, not increase them.
Therefore it would not cause old code that compiled to suddenly not be able to compile.

On the other hand, with the bug present, a programmer could make an innocent correct change to their code.
E.g., in the code example, there is correctly no error when A2|A3|A4 is the union, but if A4 is removed from the union, an incorrect error will appear, causing the code to be unable to compile.

> ... the implementation is written cleverly and for speed.

Can you give a pointer to the code?

**Comment by sandersn** [comment]
In `isRelatedTo` in checker.ts: 

```ts
const isPerformingExcessPropertyChecks = !(intersectionState & IntersectionState.Target) && (isObjectLiteralType(source) && getObjectFlags(source) & ObjectFlags.FreshLiteral);
if (isPerformingExcessPropertyChecks) {
    if (hasExcessProperties(source as FreshObjectLiteralType, target, reportErrors)) {
        if (reportErrors) {
            reportRelationError(headMessage, source, originalTarget.aliasSymbol ? originalTarget : target);
        }
        return Ternary.False;
    }
}
```

The maintenance of intersectionState and ObjectFlags.FreshLiteral is distributed across isRelatedTo. That's the part that's likely to cause problems, since the interactions of various states are confusing.

**Comment by craigphicks** [comment]
@sandersn - That is very helpful.  It seems you are saying that the problem is in the computation of `isPerformingExcessPropertyChecks`.  Could it not be a problem in the function `hasExcessProperties`? 

Aside:  To explain why I appreciate your labeling this issue as a bug, let me quote from [a reputable source on runtime type checking in typescript](https://learning-notes.mistermicheels.com/javascript/typescript/runtime-type-checking#strictness-of-runtime-checking) 

> Strictness of runtime checking
> 1. Needs to be at least as strict as compile-time checking (otherwise, we lose the guarantees that the compile-time checking provides)
> 2. Can be more strict if desired ...

Let us suppose a runtime type checking of a Union object.  The obvious baseline specification would be to mirror. as closely as reasonably possible, the excess property checking of compile time type checking.  (Stricter than baseline being optional).  

Therefore labeling this issue as a bug has helped establish a reasonable upper bound on the strictness of the definition of union excess property compile type checking.  *Would the following be an accurate description of an upper bound?*  

**An upper bound on the rules for a set of properties to satisfy a union type (|) is**
  1. **must contain all required properties of at least one union member**
  2. **may not contain additional properties that do not belong to any union member**

Note that an upper bound is not unique.  Cases which violate the the upper bound but do not result in compile time errors are allowed.  However, cases which do not violate the upper bound but become compile time errors should be called bugs (e.g., the case in this bug report). 

Please note this question is not a repeat of #44871 (now closed) - because that question was not posed in terms of an upper bound.  An upper bound definition may be a more practical formal declaration of what is presently an implicit understanding.  (Implicit understandings can easily become misunderstandings, that's why a formal declaration is preferred.)

**Comment by craigphicks** [comment]
@sandersn 
@RyanCavanaugh 
Pull with fix  - #49004
